server/internal/gateway/proxy.go

```diff
@@ -127,52 +127,90 @@ func NewToolProxy(
 	}
 }
 
-func (itp *ToolProxy) Do(
+func (tp *ToolProxy) Do(
 	ctx context.Context,
 	w http.ResponseWriter,
 	requestBody io.Reader,
 	env *CaseInsensitiveEnv,
-	gatewayTool *Tool,
-) error {
-	if !gatewayTool.IsHTTP() {
-		return fmt.Errorf("tool type not supported: %s", gatewayTool.Kind())
-	}
-
-	tool := gatewayTool.HTTPTool
-
-	ctx, span := itp.tracer.Start(ctx, "proxyToolCall", trace.WithAttributes(
-		attr.ToolName(tool.Name),
-		attr.ToolID(tool.ID),
-		attr.ProjectID(tool.ProjectID),
-		attr.DeploymentID(tool.DeploymentID),
-		attr.ToolCallSource(string(itp.source)),
-		attr.HTTPRoute(tool.Path), // this is just from the raw OpenAPI spec. It is not a path with any parameters filled in, so not identifiable.
+	plan *ToolCallPlan,
+) (err error) {
+	ctx, span := tp.tracer.Start(ctx, "gateway.toolCall", trace.WithAttributes(
+		attr.ToolName(plan.Descriptor.Name),
+		attr.ToolID(plan.Descriptor.ID),
+		attr.ProjectID(plan.Descriptor.ProjectID),
+		attr.DeploymentID(plan.Descriptor.DeploymentID),
+		attr.ToolCallSource(string(tp.source)),
 	))
-	defer span.End()
+	defer func() {
+		if err != nil {
+			span.SetStatus(codes.Error, err.Error())
+		}
+		span.End()
+	}()
 
-	logger := itp.logger.With(
-		attr.SlogProjectID(tool.ProjectID),
-		attr.SlogDeploymentID(tool.DeploymentID),
-		attr.SlogToolID(tool.ID),
-		attr.SlogToolName(tool.Name),
-		attr.SlogToolCallSource(string(itp.source)),
-		attr.SlogHTTPRoute(tool.Path), // this is just from the raw OpenAPI spec. It is not a path with any parameters filled in, so not identifiable.
+	logger := tp.logger.With(
+		attr.SlogProjectID(plan.Descriptor.ProjectID),
+		attr.SlogDeploymentID(plan.Descriptor.DeploymentID),
+		attr.SlogToolID(plan.Descriptor.ID),
+		attr.SlogToolName(plan.Descriptor.Name),
+		attr.SlogToolCallSource(string(tp.source)),
```

Comment:
having the descriptor means we can set up all the shared o11y attributes across tools